rfd/0048-war-dialler-node-tracker.md

```diff
@@ -0,0 +1,199 @@
+- Thundering herd when adding, removing, or restarting a proxy. Node agents retry connecting until they randomly get balanced to the desired proxy. The more node agents connected the worse this problem becomes.
+
+Both these issues are mitigated by changing the node agent behavior to connect to a single proxy. Ephemeral port exhaustion is no longer tied to the number of proxies and node agents no longer need to retry until they connect to a specific proxy.
+
+## Details
+
+Here's a small example comparing how proxies would interact with the node tracker:
+
+```
+	Original behaviour:
+                       ┌──────┐
+   ┌───connection─────>│proxy1│<──reverse-tunnel──┐
+   │                   └──────┘                   │
+┌──┴───┐                                        ┌─┴──┐
+│client│                                        │node│
+└──┬───┘                                        └─┬──┘
+   │                   ┌──────┐                   │
+   └───connection─────>│proxy2│<──reverse-tunnel──┘
+                       └──────┘
+	A client can reach the requested node through any proxy because the node has a reverse tunnel open to all proxies in a cluster.
+
+
+
+    New behaviour:
+                         ┌──────┐         ┌────────────┐
+   ┌─(1)─connection-1───>│proxy1├──(2)───>│node tracker│
+   │                     └──╥───┘         └────────────┘
+┌──┴───┐                    ║
+│client│               (3) tunnel
+└──┬───┘                    ║
+   │                     ┌──╨───┐                    ┌────┐
+   └────connection-2────>│proxy2│<───reverse-tunnel──┤node│
+                         └──────┘                    └────┘
+
+	Case 1 - connection 1:
+	1- A client connects to proxy-1 that can't reach the requested node.
+	2- Proxy-1 sends a lookup request to the node tracker asking which proxy is connected to the requested node. The node tracker replies with the address of proxy-2.
+	3- Proxy-1 opens a tunnel (RFD 49?) to proxy-2 and asks proxy-2 to dial the requested node.
+
+	Case 2 - connection 2:
+	nothing changes when a client connects to a proxy that has got a direct connection to the requested node.
+```
+
+### Scope
+
+The first iteration of the node tracker service will be a barebones implementation of an in memory cache represented by a simple golang map.
+
+### How it works
+
+The node tracker service runs a grpc server and exposes three endpoints to teleport proxies:
+- AddNode: which takes a `node ID`, `proxy ID` and `proxy address` as parameters among others. A Proxy will call this endpoint when it receives a heartbeat request from a node agent connected via a reverse tunnel, signaling to the node tracker that the node can be reached through this proxy, thus creating a node-to-proxy relationship.
+- RemoveNode: which takes a `node ID` as a parameter. A proxy will call this endpoint when it fails to receive heartbeat requests from a node agent after a certain threshold.
```

Comment:
I wonder if we can get rid of this rpc If the node tracker will clean up connections on its own. It might save us some issues with races when adding/removing nodes.

Reply:
I'm ok with that since we haven't focused on concurrency in this first draft. We can still reintroduce it when it's time to figure it out.